types/coin.go

```diff
@@ -703,28 +703,35 @@ func (coins Coins) AmountOf(denom string) Int {
 // AmountOfNoDenomValidation returns the amount of a denom from coins
 // without validating the denomination.
 func (coins Coins) AmountOfNoDenomValidation(denom string) Int {
+	_, c := coins.Find(denom)
+	return c.Amount
+}
+
+// Find returns true and coin if the denom exists in coins. Otherwise it returns false
+// and a zero coin. Uses binary search.
+// CONTRACT: coins must be valid (sorted).
+func (coins Coins) Find(denom string) (bool, Coin) {
 	switch len(coins) {
 	case 0:
-		return ZeroInt()
+		return false, Coin{"", ZeroInt()}
 
 	case 1:
 		coin := coins[0]
 		if coin.Denom == denom {
-			return coin.Amount
+			return true, coin
 		}
-		return ZeroInt()
+		return false, Coin{"", ZeroInt()}
```

Comment:
Why not just the empty `Coin{}` literal? The caller knows based of the bool that it's found or not.

Reply:
`Coin{}` has a nil amount. `ok=false` is not an error.

Reply:
I'm not following. You'd never use the `coin` if `ok=false`, so I don't think it matters if `coin` actually contains and we shouldn't allocate a `ZeroInt()` when we don't have to IMO

Reply:
`AmountOfNoDenomValidation` and `AmountOf` uses it. 
I agree, that it creates an allocation which is not always needed.  If you really think it's better to avoid that allocation, then I will update. lmk

Reply:
OK, updated